src/Adapter/MSTest.TestAdapter/Helpers/FixtureMethodRunner.cs

```diff
@@ -15,21 +16,76 @@ namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Helpers;
 internal static class FixtureMethodRunner
 {
     internal static TestFailedException? RunWithTimeoutAndCancellation(
-        Action action, CancellationTokenSource cancellationTokenSource, int? timeout, MethodInfo methodInfo,
+        Action action, CancellationTokenSource cancellationTokenSource, TimeoutInfo? timeoutInfo, MethodInfo methodInfo,
         IExecutionContextScope executionContextScope, string methodCancelledMessageFormat, string methodTimedOutMessageFormat)
     {
         // If no timeout is specified, we can run the action directly. This avoids any overhead of creating a task/thread and
         // ensures that the execution context is preserved (as we run the action on the current thread).
-        if (timeout is null)
+        if (timeoutInfo is null)
         {
             action();
             return null;
         }
 
+        if (timeoutInfo.Value.CooperativeCancellation)
+        {
+            return RunWithCooperativeCancellation(action, cancellationTokenSource, timeoutInfo.Value, methodInfo, methodCancelledMessageFormat, methodTimedOutMessageFormat);
```

Comment:
should this wrap RunWithTimeoutAndCancellationWithSTAThread and RunWithTimeoutAndCancellationWithThreadPool ?

Reply:
WDYM? I don't need to start a task/thread because I'll run the action synchronously as I don't need to "unobserve" it.

Reply:
so we're already in the correct thread context at this point?

Reply:
Yes! We are always in the right context before joining this method. We needed to do a lot of extra work (starting a different thread/task + moving the execution context only because we want to be able to stop observing the execution).

Reply:
But that's a good point, let me update the STA tests to confirm that!

Reply:
kk